File: app/src/main/java/com/example/flexifitapp/SignupActivity.java
package com.example.flexifitapp;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Patterns;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.content.ContextCompat;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.*;

import java.util.HashMap;
import java.util.Map;

public class SignupActivity extends AppCompatActivity {

    private TextInputEditText etName, etAddress, etEmail, etUsername, etPassword;
    private TextInputLayout tilPassword;
    private MaterialButton signupButton;
    private TextView loginRedirectText;

    private static final String PREF_NAME = "flexifit_prefs";
    private static final String KEY_DARK_MODE = "dark_mode";

    private FirebaseAuth mAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        // 🌙 Load theme
        SharedPreferences prefs = getSharedPreferences(PREF_NAME, MODE_PRIVATE);
        boolean isDark = prefs.getBoolean(KEY_DARK_MODE, false);
        AppCompatDelegate.setDefaultNightMode(
                isDark ? AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO
        );

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);

        mAuth = FirebaseAuth.getInstance();

        // Toolbar
        MaterialToolbar toolbar = findViewById(R.id.topAppBar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(v -> finish());

        // Views
        etName     = findViewById(R.id.signup_name);
        etAddress  = findViewById(R.id.signup_address);
        etEmail    = findViewById(R.id.signup_email);
        etUsername = findViewById(R.id.signup_username);
        etPassword = findViewById(R.id.signup_password);
        tilPassword = findViewById(R.id.tilpass);

        signupButton      = findViewById(R.id.signup_button);
        loginRedirectText = findViewById(R.id.loginRedirectText);

        signupButton.setOnClickListener(v -> {
            if (!validateName() | !validateAddress() | !validateEmail()
                    | !validateUsername() | !validatePassword()) {
                return;
            }
            checkIfUsernameExists();
        });

        loginRedirectText.setOnClickListener(v -> {
            startActivity(new Intent(SignupActivity.this, LoginActivity.class));
            finish();
        });
    }

    // ------------------------------------------------------------
    // 🔵 0. CHECK USERNAME DUPLICATION BEFORE SIGNUP
    // ------------------------------------------------------------
    private void checkIfUsernameExists() {
        String username = safeText(etUsername);

        DatabaseReference ref = FirebaseDatabase.getInstance()
                .getReference("users");

        Query checkUser = ref.orderByChild("username").equalTo(username);

        signupButton.setEnabled(false);

        checkUser.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {

                if (snapshot.exists()) {
                    signupButton.setEnabled(true);
                    etUsername.setError("Username already taken");
                    etUsername.requestFocus();
                    return;
                }

                // WALA pang gumagamit ng username → proceed with account creation
                createAccount();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                signupButton.setEnabled(true);
                Toast.makeText(SignupActivity.this,
                        "Database error: " + error.getMessage(),
                        Toast.LENGTH_SHORT).show();
            }
        });
    }

    // ------------------------------------------------------------
    // 🔵 1. CREATE ACCOUNT WITH FIREBASE AUTH
    // ------------------------------------------------------------
    private void createAccount() {
        String name     = safeText(etName);
        String address  = safeText(etAddress);
        String email    = safeText(etEmail);
        String username = safeText(etUsername);
        String password = safeText(etPassword);

        // 1️⃣ Create Firebase Auth account
        mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(task -> {

                    signupButton.setEnabled(true);

                    if (!task.isSuccessful()) {
                        Toast.makeText(
                                SignupActivity.this,
                                "Sign up failed: " +
                                        (task.getException() != null ? task.getException().getMessage() : ""),
                                Toast.LENGTH_LONG
                        ).show();
                        return;
                    }

                    FirebaseUser user = mAuth.getCurrentUser();
                    if (user == null) {
                        Toast.makeText(SignupActivity.this,
                                "User not available. Try again.",
                                Toast.LENGTH_SHORT).show();
                        return;
                    }

                    // 2️⃣ Send email verification WITH listener
                    user.sendEmailVerification()
                            .addOnCompleteListener(verifyTask -> {

                                if (!verifyTask.isSuccessful()) {
                                    // ❌ Hindi na-send ang email
                                    Toast.makeText(
                                            SignupActivity.this,
                                            "Failed to send verification email: " +
                                                    (verifyTask.getException() != null
                                                            ? verifyTask.getException().getMessage()
                                                            : ""),
                                            Toast.LENGTH_LONG
                                    ).show();
                                    signupButton.setEnabled(true);
                                    return; // huwag ituloy pag wala talagang email na na-send
                                }

                                // ✅ Umabot dito = verification email SENT successfully
                                String uid = user.getUid();

                                // 3️⃣ Save profile sa /users/{uid}
                                Map<String, Object> userData = new HashMap<>();
                                userData.put("uid", uid);
                                userData.put("name", name);
                                userData.put("address", address);
                                userData.put("email", email);
                                userData.put("username", username);
                                userData.put("surveyCompleted", false);
                                userData.put("createdAt", System.currentTimeMillis());

                                FirebaseDatabase.getInstance()
                                        .getReference("users")
                                        .child(uid)
                                        .setValue(userData)
                                        .addOnCompleteListener(saveTask -> {
                                            if (saveTask.isSuccessful()) {
                                                Toast.makeText(
                                                        SignupActivity.this,
                                                        "Account created! Verification link sent to your email.",
                                                        Toast.LENGTH_LONG
                                                ).show();

                                                Intent intent = new Intent(SignupActivity.this, EmailVerificationActivity.class);
                                                intent.putExtra("email", email);
                                                startActivity(intent);
                                                finish();

                                            } else {
                                                Toast.makeText(SignupActivity.this,
                                                        "Failed to save profile. Try again.",
                                                        Toast.LENGTH_SHORT).show();
                                            }
                                        });
                            });
                });
    }

    // ------------------------------------------------------------
    // VALIDATIONS
    // ------------------------------------------------------------

    private boolean validateName() {
        String val = safeText(etName);
        if (val.isEmpty()) {
            etName.setError("Name cannot be empty");
            return false;
        } else if (val.length() < 2) {
            etName.setError("Name is too short");
            return false;
        } else {
            etName.setError(null);
            return true;
        }
    }

    private boolean validateAddress() {
        String val = safeText(etAddress);
        if (val.isEmpty()) {
            etAddress.setError("Address cannot be empty");
            return false;
        } else {
            etAddress.setError(null);
            return true;
        }
    }

    private boolean validateEmail() {
        String val = safeText(etEmail);
        if (val.isEmpty()) {
            etEmail.setError("Email cannot be empty");
            return false;
        } else if (!Patterns.EMAIL_ADDRESS.matcher(val).matches()) {
            etEmail.setError("Invalid email format");
            return false;
        } else {
            etEmail.setError(null);
            return true;
        }
    }

    private boolean validateUsername() {
        String val = safeText(etUsername);
        if (val.isEmpty()) {
            etUsername.setError("Username cannot be empty");
            return false;
        } else if (val.contains(" ")) {
            etUsername.setError("Username cannot have spaces");
            return false;
        } else if (val.length() < 2) {
            etUsername.setError("Username must be at least 2 characters");
            return false;
        } else {
            etUsername.setError(null);
            return true;
        }
    }

    private boolean validatePassword() {
        String val = safeText(etPassword);
        if (val.isEmpty()) {
            tilPassword.setError("Password cannot be empty");
            return false;
        } else if (val.length() < 6) {
            tilPassword.setError("Password must be at least 6 characters");
            return false;
        } else {
            tilPassword.setError(null);
            return true;
        }
    }

    private String safeText(TextInputEditText et) {
        return et.getText() != null ? et.getText().toString().trim() : "";
    }

    // 🌙 Theme menu
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_theme_switcher, menu);

        MenuItem item = menu.findItem(R.id.action_toggle_theme);
        boolean isDark = getSharedPreferences(PREF_NAME, MODE_PRIVATE)
                .getBoolean(KEY_DARK_MODE, false);

        item.setIcon(ContextCompat.getDrawable(
                this,
                isDark ? R.drawable.ic_sun : R.drawable.ic_moon
        ));

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_toggle_theme) {
            SharedPreferences prefs = getSharedPreferences(PREF_NAME, MODE_PRIVATE);
            boolean current = prefs.getBoolean(KEY_DARK_MODE, false);
            boolean newMode = !current;

            prefs.edit().putBoolean(KEY_DARK_MODE, newMode).apply();

            AppCompatDelegate.setDefaultNightMode(
                    newMode ? AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO
            );

            recreate();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/flexifitapp/SignupActivity.kt
package com.example.flexifitapp

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Patterns
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat
import androidx.credentials.CredentialManager
import androidx.credentials.GetCredentialRequest
import androidx.lifecycle.lifecycleScope
import com.example.flexifitapp.auth.LoginRequest
import com.example.flexifitapp.googleauth.CreateUsernameActivity
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.android.libraries.identity.googleid.GoogleIdTokenParsingException
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import kotlinx.coroutines.launch

class SignupActivity : AppCompatActivity() {

    private lateinit var etName: TextInputEditText
    private lateinit var etEmail: TextInputEditText
    private lateinit var etUsername: TextInputEditText
    private lateinit var etPassword: TextInputEditText
    private lateinit var tilPassword: TextInputLayout
    private lateinit var btnSignup: MaterialButton
    private lateinit var btnGoogleSignup: MaterialButton
    private lateinit var cbTermsSignup: CheckBox
    private lateinit var tvLoginRedirect: TextView

    private lateinit var auth: FirebaseAuth
    private lateinit var prefs: SharedPreferences
    private lateinit var credentialManager: CredentialManager

    private val PREF_NAME = "flexifit_prefs"
    private val KEY_DARK_MODE = "dark_mode"

    private val KEY_PENDING_NAME = "pending_name"
    private val KEY_PENDING_USERNAME = "pending_username"
    private val KEY_PENDING_EMAIL = "pending_email"
    private val KEY_PENDING_CREATED_AT = "pending_created_at"

    private var isSubmitting = false

    override fun onCreate(savedInstanceState: Bundle?) {
        prefs = getSharedPreferences(PREF_NAME, MODE_PRIVATE)
        applyThemeFromPrefs()

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_signup)

        auth = FirebaseAuth.getInstance()
        credentialManager = CredentialManager.create(this)

        val toolbar = findViewById<MaterialToolbar>(R.id.topAppBar)
        setSupportActionBar(toolbar)
        toolbar.setNavigationOnClickListener { finish() }

        etName = findViewById(R.id.signup_name)
        etEmail = findViewById(R.id.signup_email)
        etUsername = findViewById(R.id.signup_username)
        etPassword = findViewById(R.id.signup_password)
        tilPassword = findViewById(R.id.tilpass)
        btnSignup = findViewById(R.id.signup_button)
        btnGoogleSignup = findViewById(R.id.btnGoogleSignup)
        cbTermsSignup = findViewById(R.id.cbTermsSignup)
        tvLoginRedirect = findViewById(R.id.loginRedirectText)

        intent.getStringExtra("email")?.let { etEmail.setText(it) }

        btnSignup.setOnClickListener {
            if (isSubmitting) return@setOnClickListener

            clearErrors()

            if (!cbTermsSignup.isChecked) {
                Toast.makeText(
                    this,
                    "Please accept the Terms of Service first.",
                    Toast.LENGTH_SHORT
                ).show()
                return@setOnClickListener
            }

            if (!validateAll()) return@setOnClickListener
            createAccount()
        }

        btnGoogleSignup.setOnClickListener {
            if (isSubmitting) return@setOnClickListener

            clearErrors()

            if (!cbTermsSignup.isChecked) {
                Toast.makeText(
                    this,
                    "Please accept the Terms of Service first.",
                    Toast.LENGTH_SHORT
                ).show()
                return@setOnClickListener
            }

            startGoogleSignUp()
        }

        tvLoginRedirect.setOnClickListener {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }
    }

    private fun createAccount() {
        val name = safeText(etName)
        val email = safeText(etEmail)
        val username = safeText(etUsername)
        val password = safeText(etPassword)

        setLoading(true)
        savePendingSignup(name, username, email)

        auth.createUserWithEmailAndPassword(email, password)
            .addOnCompleteListener { task ->
                if (!task.isSuccessful) {
                    setLoading(false)
                    handleSignupError(task.exception)
                    return@addOnCompleteListener
                }

                val user = auth.currentUser
                if (user == null) {
                    setLoading(false)
                    Toast.makeText(this, "Signup failed. Please try again.", Toast.LENGTH_LONG)
                        .show()
                    return@addOnCompleteListener
                }

                sendVerificationAndRedirect(user, email)
            }
    }

    private fun sendVerificationAndRedirect(user: FirebaseUser, email: String) {
        user.sendEmailVerification()
            .addOnCompleteListener { verifyTask ->
                // Check if the activity is still alive
                if (isFinishing || isDestroyed) return@addOnCompleteListener

                setLoading(false)

                if (!verifyTask.isSuccessful) {
                    Toast.makeText(
                        this,
                        "Failed to send verification email: ${verifyTask.exception?.message ?: "Unknown error"}",
                        Toast.LENGTH_LONG
                    ).show()
                    return@addOnCompleteListener
                }

                auth.signOut()

                Toast.makeText(
                    this,
                    "Account created! Please verify via the link sent to your email.",
                    Toast.LENGTH_LONG
                ).show()

                val intent = Intent(this, EmailVerificationActivity::class.java).apply {
                    putExtra("email", email)
                    addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                }
                startActivity(intent)
                finish()
            }
    }

    private fun startGoogleSignUp() {
        val googleIdOption = GetGoogleIdOption.Builder()
            .setFilterByAuthorizedAccounts(false)
            .setServerClientId(getString(R.string.default_web_client_id))
            .setAutoSelectEnabled(false)
            .build()

        val request = GetCredentialRequest.Builder()
            .addCredentialOption(googleIdOption)
            .build()

        setLoading(true)

        lifecycleScope.launch {
            try {
                val result = credentialManager.getCredential(
                    context = this@SignupActivity,
                    request = request
                )

                val credential = result.credential
                val googleIdTokenCredential = GoogleIdTokenCredential.createFrom(credential.data)
                val googleIdToken = googleIdTokenCredential.idToken

                firebaseAuthWithGoogle(googleIdToken)

            } catch (e: GoogleIdTokenParsingException) {
                setLoading(false)
                Toast.makeText(
                    this@SignupActivity,
                    "Invalid Google credential.",
                    Toast.LENGTH_LONG
                ).show()
            } catch (e: Exception) {
                setLoading(false)
                Toast.makeText(
                    this@SignupActivity,
                    "Google sign up failed: ${e.message ?: "Unknown error"}",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    private fun firebaseAuthWithGoogle(idToken: String) {
        val firebaseCredential = GoogleAuthProvider.getCredential(idToken, null)

        auth.signInWithCredential(firebaseCredential)
            .addOnCompleteListener(this) { task ->
                if (!task.isSuccessful) {
                    setLoading(false)
                    Toast.makeText(
                        this,
                        "Firebase Google auth failed: ${task.exception?.message ?: "Unknown error"}",
                        Toast.LENGTH_LONG
                    ).show()
                    return@addOnCompleteListener
                }

                val user = auth.currentUser
                if (user == null) {
                    setLoading(false)
                    Toast.makeText(this, "Google user not found.", Toast.LENGTH_LONG).show()
                    return@addOnCompleteListener
                }

                checkIfGoogleUserExists(user)
            }
    }

    private fun checkIfGoogleUserExists(user: FirebaseUser) {
        user.getIdToken(true).addOnCompleteListener { idTask ->
            val firebaseToken = idTask.result?.token

            if (!idTask.isSuccessful || firebaseToken.isNullOrBlank()) {
                setLoading(false)
                Toast.makeText(
                    this,
                    "Token error: ${idTask.exception?.message ?: "Unknown error"}",
                    Toast.LENGTH_LONG
                ).show()
                return@addOnCompleteListener
            }

            lifecycleScope.launch {
                try {
                    val api = ApiClient.get(this@SignupActivity)
                        .create(ApiService::class.java)

                    val loginReq = LoginRequest(
                        firebaseIdToken = firebaseToken,
                        fcmToken = null
                    )

                    val res = api.login(loginReq)

                    if (res.isSuccessful && res.body() != null) {
                        val authBody = res.body()!!

                        // Sa loob ng checkIfGoogleUserExists...
                        UserPrefs.saveAuth(
                            this@SignupActivity,
                            authBody.token,
                            authBody.userId,
                            authBody.role,
                            authBody.status,
                            authBody.isVerified,
                            authBody.name,    // Ika-anim na parameter
                            authBody.photoUrl // Ika-pitong parameter
                        )
                        val bootRes = api.bootstrap()

                        if (!bootRes.isSuccessful || bootRes.body() == null) {
                            setLoading(false)
                            Toast.makeText(
                                this@SignupActivity,
                                "Bootstrap failed: ${bootRes.code()}",
                                Toast.LENGTH_LONG
                            ).show()
                            return@launch
                        }

// Hanapin mo itong line sa lifecycleScope.launch:
                        val body = bootRes.body()!!

                        if (body.userId != null) {
                            UserPrefs.putInt(this@SignupActivity, UserPrefs.KEY_USER_ID, body.userId)
                        }

// DAPAT MAG-MATCH SA BOOTSTRAPRESPONSE MODEL NATIN
                        if (body.profileComplete) { // <--- Gamitin ang variable name na nasa Model mo
                            goToMain()
                        } else {
                            goToOnboard()
                        }
                    } else {
                        openCreateUsername(user, firebaseToken)
                    }

                } catch (e: Exception) {
                    setLoading(false)
                    Toast.makeText(
                        this@SignupActivity,
                        e.message ?: "Login check failed",
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
        }
    }

    private fun openCreateUsername(user: FirebaseUser, firebaseToken: String) {
        val intent = Intent(this, CreateUsernameActivity::class.java).apply {
            putExtra("firebaseToken", firebaseToken)
            putExtra("name", user.displayName.orEmpty())
            putExtra("email", user.email.orEmpty())
        }
        startActivity(intent)
        finish()
    }

    private fun goToMain() {
        val intent = Intent(this, MainActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        startActivity(intent)
        finish()
    }

    private fun goToOnboard() {
        val intent = Intent(this, OnboardingActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        startActivity(intent)
        finish()
    }

    private fun handleSignupError(ex: Exception?) {
        when (ex) {
            is FirebaseAuthUserCollisionException -> {
                etEmail.error = "This email is already registered."
                Toast.makeText(this, "Email already in use. Try logging in.", Toast.LENGTH_LONG)
                    .show()
            }

            else -> {
                Toast.makeText(
                    this,
                    "Sign up failed: ${ex?.message ?: "Unknown error"}",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    private fun savePendingSignup(name: String, username: String, email: String) {
        prefs.edit()
            .putString(KEY_PENDING_NAME, name)
            .putString(KEY_PENDING_USERNAME, username)
            .putString(KEY_PENDING_EMAIL, email)
            .putLong(KEY_PENDING_CREATED_AT, System.currentTimeMillis())
            .apply()
    }

    private fun validateAll(): Boolean {
        val okName = validateName()
        val okEmail = validateEmail()
        val okUser = validateUsername()
        val okPass = validatePassword()
        return okName && okEmail && okUser && okPass
    }

    private fun validateName(): Boolean {
        val v = safeText(etName)
        return when {
            v.isEmpty() -> {
                etName.error = "Name cannot be empty"
                false
            }

            v.length < 2 -> {
                etName.error = "Name is too short"
                false
            }

            else -> true
        }
    }

    private fun validateEmail(): Boolean {
        val v = safeText(etEmail)
        return when {
            v.isEmpty() -> {
                etEmail.error = "Email cannot be empty"
                false
            }

            !Patterns.EMAIL_ADDRESS.matcher(v).matches() -> {
                etEmail.error = "Invalid email format"
                false
            }

            else -> true
        }
    }

    private fun validateUsername(): Boolean {
        val v = safeText(etUsername)
        return when {
            v.isEmpty() -> {
                etUsername.error = "Username cannot be empty"
                false
            }

            v.contains(" ") -> {
                etUsername.error = "Username cannot have spaces"
                false
            }

            v.length < 2 -> {
                etUsername.error = "Username must be at least 2 characters"
                false
            }

            else -> true
        }
    }

    private fun validatePassword(): Boolean {
        val v = safeText(etPassword)
        return when {
            v.isEmpty() -> {
                tilPassword.error = "Password cannot be empty"
                false
            }

            v.length < 6 -> {
                tilPassword.error = "Password must be at least 6 characters"
                false
            }

            else -> {
                tilPassword.error = null
                true
            }
        }
    }

    private fun clearErrors() {
        etName.error = null
        etEmail.error = null
        etUsername.error = null
        tilPassword.error = null
    }

    private fun safeText(et: TextInputEditText): String {
        return et.text?.toString()?.trim().orEmpty()
    }

    private fun setLoading(loading: Boolean) {
        isSubmitting = loading

        btnSignup.isEnabled = !loading
        btnGoogleSignup.isEnabled = !loading
        cbTermsSignup.isEnabled = !loading
        tvLoginRedirect.isEnabled = !loading

        btnSignup.text = if (loading) "Creating..." else "Sign Up"
        btnGoogleSignup.text = if (loading) "Please wait..." else "Sign up with Google"

        setEnabled(etName, !loading)
        setEnabled(etEmail, !loading)
        setEnabled(etUsername, !loading)
        setEnabled(etPassword, !loading)

        val overlay = findViewById<View?>(R.id.loadingOverlay)
        overlay?.visibility = if (loading) View.VISIBLE else View.GONE
    }

    private fun setEnabled(view: View, enabled: Boolean) {
        view.isEnabled = enabled
        view.isClickable = enabled
        view.isFocusable = enabled
        view.isFocusableInTouchMode = enabled
    }

    private fun applyThemeFromPrefs() {
        val isDark = prefs.getBoolean(KEY_DARK_MODE, false)
        AppCompatDelegate.setDefaultNightMode(
            if (isDark) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO
        )
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_theme_switcher, menu)
        val item = menu.findItem(R.id.action_toggle_theme)
        val isDark = prefs.getBoolean(KEY_DARK_MODE, false)
        item.icon = ContextCompat.getDrawable(
            this,
            if (isDark) R.drawable.ic_sun else R.drawable.ic_moon
        )
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_toggle_theme) {
            val current = prefs.getBoolean(KEY_DARK_MODE, false)
            val newMode = !current
            prefs.edit().putBoolean(KEY_DARK_MODE, newMode).apply()
            AppCompatDelegate.setDefaultNightMode(
                if (newMode) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO
            )
            recreate()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}